include/antler/project/net_utils.hpp

```diff
@@ -0,0 +1,190 @@
+         try {
+            (void)request(url);
+            return true;
+         } catch(...) {
+            return false;
+         }
+      }
+
+      CURL* curl_obj;
+   };
+
+   // object to encapsulate github api
+   struct github {
+      inline github() {
+         has_gh_app = system::execute_quiet("gh", {"--version"}) == 0;
+         system::debug_log("Found and using github app!");
+
+         if (!has_gh_app) {
+            bearer_token = sender.request("https://ghcr.io/token?service=registry.docker.io&scope=repository:AntelopeIO/experimental-binaries:pull");
+            bearer_token = bearer_token.substr(bearer_token.find(":\"")+2);
+            bearer_token = bearer_token.substr(0, bearer_token.size()-3);
+            system::debug_log("Using default bearer token: {0}", bearer_token);
+         } else {
+            bearer_token = system::execute("gh", {"auth", "token"});
+            system::debug_log("Using github app bearer token: {0}", bearer_token);
+         }
+      }
+
+      std::string request(std::string_view org, std::string_view repo) {
+         std::string url = "https://api.github.com/repos/";
+         url += std::string(org) + std::string("/");
+         url += repo;
+         auto s = sender.authenticated_request(bearer_token, url.c_str());
+         return s;
+      }
+
+      inline std::string request(std::string_view s) { return request(get_org(s), get_repo(s)); }
+
+      inline std::string get_default_branch(std::string_view org, std::string_view repo) {
+         auto js = nlohmann::json::parse(request(org, repo));
+         auto def_branch = js["default_branch"];
+         ANTLER_CHECK(!def_branch.is_null(), "github api failure, default branch not found");
+         return def_branch.get<std::string>();
+      }
+
+      inline bool is_reachable(std::string_view s) {
+         try {
+            auto js = nlohmann::json::parse(request(s));
+            auto msg = js["message"];
+            if (!msg.is_null())
+               if (msg == "Not Found")
+                  return false;
+            return true;
+         } catch(...) {
+            return false;
+         }
+      }
+
+      static std::string_view get_org(std::string_view s) {
+         auto sub = s.substr(0, s.find_last_of("/"));
+         auto pos = sub.find_last_of("/");
+         return pos == std::string_view::npos ? 
+               sub :
+               sub.substr(pos+1);
+      }
+
+      static std::string_view get_repo(std::string_view s) {
+         auto pos = s.find_last_of("/");
+         if (pos == std::string_view::npos)
+            return "";
+         return s.substr(pos+1);
+      }
+
+      static inline bool is_shorthand(std::string_view s) {
+         auto sub = s.substr(0, s.find_last_of("/"));
+         return sub.size() != s.size() && sub.find_last_of("/") == std::string_view::npos;
+      }
+
+      curl sender;
+      std::string bearer_token = "";
+      bool has_gh_app          = false;
```

Comment:
I don't believe these should be used outside of the scope of this struct/class so it makes sense to protect them:
```suggestion
   private:
      curl sender;
      std::string bearer_token = "";
      bool has_gh_app          = false;
```

Reply:
Fixed in 03260ee4